cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/service/DraftHandler.java

```diff
@@ -0,0 +1,191 @@
+import javax.annotation.Nullable;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+/**
+ * Handler of drafts
+ */
+public class DraftHandler extends AbstractSystemHttpServiceHandler {
+  private static final Gson GSON = new GsonBuilder()
+    .setPrettyPrinting()
+    .registerTypeAdapter(DraftRequest.class, new DraftRequestDeserializer())
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .create();
+
+  private final Metrics metrics;
+
+  private DraftService draftService;
+
+  public DraftHandler(Metrics metrics) {
+    this.metrics = metrics;
+  }
+
+  @Override
+  public void initialize(SystemHttpServiceContext context) throws Exception {
+    super.initialize(context);
+    this.draftService = new DraftService(context, this.metrics);
+  }
+
+  /**
+   * Returns a list of drafts associated with a namespace and the current user
+   */
+  @GET
+  @Path("v1/contexts/{context}/drafts")
+  public void listDrafts(HttpServiceRequest request, HttpServiceResponder responder,
+                         @PathParam("context") String namespaceName,
+                         @QueryParam("includeConfig") @DefaultValue("false") boolean includeConfig,
+                         @QueryParam("sortBy") @DefaultValue("name") String sortBy,
+                         @QueryParam("sortOrder") @DefaultValue("ASC") String sortOrder,
+                         @QueryParam("filter") @Nullable String filter) {
+
+    respond(namespaceName, responder, (namespace) -> {
+      SortRequest sortRequest = new SortRequest(sortBy, sortOrder);
+      responder.sendJson(draftService.listDrafts(namespace, request.getUserId(), includeConfig, sortRequest, filter));
+    });
+  }
+
+  /**
+   * Gets the details of a draft
+   */
+  @GET
+  @Path("v1/contexts/{context}/drafts/{draft}")
+  public void getDraft(HttpServiceRequest request, HttpServiceResponder responder,
+                       @PathParam("context") String namespaceName,
+                       @PathParam("draft") String draftId) {
+    respond(namespaceName, responder, (namespace) -> {
+      DraftId id = new DraftId(namespace, draftId, request.getUserId());
+      responder.sendJson(draftService.getDraft(id));
+    });
+  }
+
+  /**
+   * Creates or updates a draft
+   */
+  @PUT
+  @Path("v1/contexts/{context}/drafts/{draft}")
+  public void putDraft(HttpServiceRequest request, HttpServiceResponder responder,
+                       @PathParam("context") String namespaceName,
+                       @PathParam("draft") String draftId) {
+
+    respond(namespaceName, responder, (namespace) -> {
+      DraftRequest<ETLConfig> draftRequest;
+      try {
+        draftRequest = GSON
+          .fromJson(StandardCharsets.UTF_8.decode(request.getContent()).toString(), DraftRequest.class);
+      } catch (JsonSyntaxException e) {
```

Comment:
Why not move it to the `respond` method to handle as well?

Reply:
This draftRequest object is only needed for the putDraft endpoint, adding this to the respond would mean that it would try to parse this for every endpoint right?